### ISSUE TITLE
midi : midi out only : using instrument as sequencer for external app

### ISSUE THREAD

**Issue by thijz** [hydrogen-music/hydrogen#63]
currently if you want to use a H2 instrument as a pure sequencer (ie the notes in the pattern are only sent out as MIDI notes. H2 does not generate any sound) you MUST still load a long empty wav file into that instrument.
if you do not do this the MIDI notes will have a duration of 0. as a consequence the external app will not play anything.

the best option would be that H2 always sends out MIDI notes regardless of the duration of the loaded WAV file
an alternative would be to add a 'MIDI out only' tickbox on the General tab of the Instrument editor that automatically loads a 'very long' empty WAV into that instrument

this second option however is only a quick workaround

comment by Chris Mennie :
The purpose of having an associated sound file (which admittedly is a huge hack) is to tell Hydrogen how long to hold the note before a note off message is sent. So it's not just a case of specifying a "very long" duration of sound file because different drum sounds require different play lengths.

As far as the code goes, I believe it'd be straight forward to define a default note length to use when there's no associated sound sample. However, this doesn't really solve the real issue (though it does go much farther than what's currently in place). I don't know how it should be presented, but the user needs to be able to specify the note length (via the UI? via midi input?).

**Comment by thijz** [comment]
similar to #57 and #195

**Comment by rghvdberg** [comment]
I've tried to use H2 as a sequencer for synthesized drums.
It failed horribly because the Note On / Off messages overlapped.
The Note On message was sent before the Note Off message.

**Comment by Yaoir** [comment]
I looked in the MIDI 1.0 specification to see if there was a recommended way to handle note off events for one-shot sounds (such as drum hits), and couldn't find anything. But it did say that every note on should be followed by a note off.

I think the only sensible way to handle this is for Hydrogen to emit a note off immediately following the note on, and assume the receiver will ignore the note off.  Sending the note off when the internal sample finishes playing is simply wrong. It makes no sense, and according to rghvdberg can cause problems.

In any case, I suggest that the subject of controlling MIDI devices from Hydrogen be written up in the manual. The way Hydrogen currently acts definitely needs to be documented there because it's very unintuitive, and without proper documentation, anyone who creates a new instrument with the sole purpose of sending MIDI to a synth will have problems.

**Comment by m-grabner** [comment]
I had the same problem and opened a separate tracker issue (#533), elpescado redirected me here. With the code modified as posted under #533, I get a "note off - note on - note off" sequence in a very short time (some milliseconds) for each note in Hydrogen, and when connected to LinuxSampler, the notes are played as expected (i.e., "note off" messages are ignored). With unmodified Hydrogen-0.9.7, only "not off" messages are produced, which certainly don't trigger any sound output in LinuxSampler.

BTW, not everyone agrees with a "note off" being required after each "note on". There is a MIDI tutorial (http://www.music-software-development.com/midi-tutorial.html) where percussion instruments are explicitly excluded from this rule. Why not simply create a checkbox "Generate note-off" near the checkbox "Ignore note-off" and let the user decide? Variable-length note are probably harder to implement, and not really useful in a drum machine.

Markus

**Comment by thijz** [comment]
@Yaoir : very interesting observation about the different behavior when recording the note using a midi device and entering it manually.
When you did this test, did you have a sample loaded in the instrument you were recording the midi notes for ?  or was it an empty instrument ?

@m-grabner : you are correct about not requiring a 'note off' for percussion instruments, and the option to simply not send any note off messages would indeed be a nice addition.  But keep in mind that quite a few people (including me :-) use H2 not only for drums, but also as a synth sequencer.  In that scenario the note off is obviously very relevant :-)

**Comment by mauser** [comment]
Hi,

i suppose that adding an "Omit NoteOff at end of sample" option should be easy to implement and would be helpful in such use cases. 

@thijz : This could be a nice method to define note length for empty samples.. I will have a look into this..

**Comment by elpescado** [comment]
@thijz there are two kinds of notes in Hydrogen. "Dots", and "rectangles" which have defined duration. That option  should affect only "dots".

**Comment by thijz** [comment]
@elpescado : good observation, so a dot would be a 'percussion note' (=only note ON) and a rectangle would be the equivalent of the traditional midi notation in most DAWs like cubase/logic : a 'note on' at the beginning of the rectangle, and a 'note off' at the end.

So this means that for the 'rectangle' the length of the sample should never affect the midi note on/off messages, right ?

**Comment by elpescado** [comment]
What's the idea behind sending note off when sample playback ends anyway? For me it doesn't make any sense, but I might be missing something.

In my opinion, there are two cases to consider.

1. Square notes - this one is clear, "note on" when note starts, "note off" when note ends
2. Dot notes - here are few possibilities
 - don't send note off at all
 - send note off when sample playback ends (prohibits using Hydrogen as sequencer only)
 - send note off instantly after note on
 - send note off after predefined interval (eg. 1/4)

We should discuss whether this should be hardcoded or configurable, and in latter case, on what level (song vs application).

I lean towards not sending note-off messages for "dot" notes. I think it's in line with MIDI spec (see @m-grabner  comment above).

**Comment by thijz** [comment]
+1 for dot=note ON _only_
all other use cases just dont make sense

**Comment by mauser** [comment]
@thijz , @elpescado :

Would you be willing to create a wiki page with the use cases and the description of the current situation concerning note off? This could be a help with the discussion for the moment and a good documentation for the future. Thijz, i remember that you also mentioned note off problems when using the Panic button.

Another thing: I suppose the midi export has also to be considered. Maybe there it makes sense to embed also note off notes at sample end? Since we don't know to what machines/synths those midi tracks will be fed to..

**Comment by thijz** [comment]
@mauser : ok, will do
@elpescado : as soon as i have something i will post it here and invite you to add more info to the page

**Comment by thijz** [comment]
wiki page : https://github.com/hydrogen-music/hydrogen/wiki/MIDI-implementation

This wiki page is work in progress, feel free to add/comment

**Comment by elpescado** [comment]
My use case is using Hydrogen as pure sequencer in tandem with external sampler, so sound is generated solely by second application. In that case, I'd like to create a drumkit, assign MIDI notes to instruments but not load any sample sounds.

However, Hydrogen refuses to play MIDI notes unless there's at leas one sample assigned to instrument.

**Comment by theGreatWhiteShark** [comment]
Alright, AFAICS there are a number of different things to be done in here:

- [ ] An option in the Preferences to omit NOTE_OFF messages for notes without user-defined note length.
- [ ] A fallback to the behavior above for notes of instruments featuring no samples at all.
- [ ] Send a NOTE_OFF message at the end of an user-defined note length (rectangle).

anyone still around for testing?

**Comment by devloop0123** [comment]
> > I still need to load a sample into an instrument before I get any midi output. Without it, any notes I draw are mute, both visually in the pattern editor as in the midi out monitor.
> > Hydrogen Drum Machine 2.0.0-pre-alpha-'1ed20759b' [Jan 9 2026]
> 
> Ah, damn. Seems I relied too much on unit tests and have not done a proper integration test. Could you try [#2270](https://github.com/hydrogen-music/hydrogen/pull/2270)? Should be fixed in there.
> 
> And thanks for testing!

Yes, now it sends midi. But note-off is sent immediately after the note-on, not respecting any custom note lengths, not respecting the setting in the midi output settings dialog (or is that not implemented?).

**Comment by devloop0123** [comment]
> Should be fixed in [064710f](https://github.com/hydrogen-music/hydrogen/commit/064710f2257a611d2d52fc7c2e4e5784c176e61e) .

Confirmed!
 
> From a user perspective: would you expect such behavior? Or actually want it?

No, with pure midi it's just confusing at best.

**Comment by devloop0123** [comment]
> > No, with pure midi it's just confusing at best.
> 
> And within the pattern editor? There Hydrogen does automatically scale the length of a note when dragging it up or down within the piano roll editor.

Horrible UX imo. Also, the scaling is only done visually. So if you want to create a 1/4th length C5 note, you'd actually have to draw a 1/16th length C5 note.
